File: src/main/java/ankol/mod/merger/merger/scr/TechlandScrFileMerger.java
package ankol.mod.merger.merger.scr;

import ankol.mod.merger.antlr.scr.TechlandScriptLexer;
import ankol.mod.merger.antlr.scr.TechlandScriptParser;
import ankol.mod.merger.core.FileMerger;
import ankol.mod.merger.core.MergerContext;
import ankol.mod.merger.exception.BusinessException;
import ankol.mod.merger.merger.ConflictRecord;
import ankol.mod.merger.merger.MergeResult;
import ankol.mod.merger.merger.scr.node.ScrContainerScriptNode;
import ankol.mod.merger.merger.scr.node.ScrFunCallScriptNode;
import ankol.mod.merger.merger.scr.node.ScrScriptNode;
import ankol.mod.merger.tools.ColorPrinter;
import ankol.mod.merger.tools.FileTree;
import ankol.mod.merger.tools.Localizations;
import ankol.mod.merger.tools.Tools;
import cn.hutool.cache.Cache;
import cn.hutool.cache.CacheUtil;
import cn.hutool.core.util.StrUtil;
import lombok.extern.slf4j.Slf4j;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.TokenStreamRewriter;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.Objects;

@Slf4j
public class TechlandScrFileMerger extends FileMerger {
    /**
     * 标记冲突项的容器
     */
    private final List<ConflictRecord> conflicts = new ArrayList<>();

    /**
     * 插入操作记录
     */
    private record InsertOperation(int tokenIndex, String content) {
    }

    private final List<InsertOperation> insertOperations = new ArrayList<>();

    /**
     * Parse 缓存，避免重复解析相同内容的文件
     * 存储 ParseResult 包含 AST 和 TokenStream
     */
    private static final Cache<String, ParseResult> PARSE_CACHE = CacheUtil.newWeakCache(60 * 1000);

    /**
     * 基准MOD（data0.pak）对应文件的语法树，用于三方对比
     */
    private ScrContainerScriptNode originalBasModRoot = null;

    /**
     * 解析结果包装类，包含AST和TokenStream
     */
    private record ParseResult(ScrContainerScriptNode astNode, CommonTokenStream tokens) {
    }

    public TechlandScrFileMerger(MergerContext context) {
        super(context);
    }

    @Override
    public MergeResult merge(FileTree file1, FileTree file2) {
        try {
            // 解析基准MOD文件（如果存在）
            if (context.getOriginalBaseModContent() != null) {
                String contentHash = Tools.computeHash(context.getOriginalBaseModContent());
                ParseResult cached = PARSE_CACHE.get(contentHash);
                if (cached != null) {
                    originalBasModRoot = cached.astNode;
                } else {
                    ParseResult result = parseContent(context.getOriginalBaseModContent());
                    originalBasModRoot = result.astNode;
                    PARSE_CACHE.put(contentHash, result);
                }
            }
            // 解析base和mod文件，保留TokenStream
            ParseResult baseResult = parseFile(file1.getFullPathName());
            ParseResult modResult = parseFile(file2.getFullPathName());
            ScrContainerScriptNode baseRoot = baseResult.astNode;
            ScrContainerScriptNode modRoot = modResult.astNode;
            // 递归对比，找到冲突项
            reduceCompare(originalBasModRoot, baseRoot, modRoot);
            //第一个mod与原版文件的对比，直接使用MOD修改的版本，不提示冲突
            if (context.isFirstModMergeWithBaseMod() && !conflicts.isEmpty()) {
                for (ConflictRecord record : conflicts) {
                    record.setUserChoice(2); // 自动选择第一个mod的版本
                }
            } else if (!conflicts.isEmpty()) {
                // 正常情况下，提示用户解决冲突
                resolveConflictsInteractively();
            }
            String mergedContent = getMergedContent(baseResult);
            return new MergeResult(mergedContent, !conflicts.isEmpty());
        } catch (Exception e) {
            log.error(StrUtil.format("Error during SCR file merge: {} Reason: {}", file1.getFileName(), e.getMessage()), e);
            throw new BusinessException("文件" + file1.getFileName() + "合并失败");
        } finally {
            //清理状态，准备下一个文件合并
            conflicts.clear();
            insertOperations.clear();
            originalBasModRoot = null;
        }
    }

    private String getMergedContent(ParseResult baseResult) {
        TokenStreamRewriter rewriter = new TokenStreamRewriter(baseResult.tokens);

        // 处理冲突节点的替换
        for (ConflictRecord record : conflicts) {
            if (record.getUserChoice() == 2) { // 用户选择了 Mod
                ScrScriptNode baseNode = record.getBaseNode();
                ScrScriptNode modNode = record.getModNode();

                // 直接使用节点中存储的token索引
                rewriter.replace(
                        baseNode.getStartTokenIndex(),
                        baseNode.getStopTokenIndex(),
                        modNode.getSourceText()
                );
            }
        }


        // 处理新增节点（插入操作）
        for (InsertOperation op : insertOperations) {
            rewriter.insertBefore(op.tokenIndex, op.content);
        }

        // 获取重写后的文本
        return rewriter.getText();
    }

    private void reduceCompare(ScrContainerScriptNode originalContainer, ScrContainerScriptNode baseContainer, ScrContainerScriptNode modContainer) {
        // 遍历 Mod 的所有子节点
        for (Map.Entry<String, ScrScriptNode> entry : modContainer.getChildren().entrySet()) {
            try {
                String signature = entry.getKey();
                ScrScriptNode modNode = entry.getValue();
                ScrScriptNode originalNode = null;
                if (originalContainer != null) {
                    originalNode = originalContainer.getChildren().get(signature);
                }
                ScrScriptNode baseNode = baseContainer.getChildren().get(signature);

                if (baseNode == null) {
                    // 新增 Base 没有这个节点 -> 插入
                    handleInsertion(baseContainer, modNode);
                } else {
                    // [存在] 检查是否冲突
                    if (baseNode instanceof ScrContainerScriptNode && modNode instanceof ScrContainerScriptNode) {
                        // 容器节点，递归进入内部对比
                        reduceCompare((ScrContainerScriptNode) originalNode, (ScrContainerScriptNode) baseNode, (ScrContainerScriptNode) modNode);
                    }
                    //函数调用，比较参数，不对比String，因为对比字符会有各种空行不规范问题
                    else if (baseNode instanceof ScrFunCallScriptNode baseFunCall && modNode instanceof ScrFunCallScriptNode modFunCall) {
                        //先检查当前待合并节点内容是否相同
                        if (!baseFunCall.getArguments().equals(modFunCall.getArguments())) {
                            //两者内容不同，检查mod节点内容与原版是否相同
                            if (!isNodeSameAsOriginalNode(originalNode, modNode)) {
                                //当前节点与原版一致，说明此处节点未变动，使用mod的内容
                                if (isNodeSameAsOriginalNode(originalNode, baseFunCall)) {
                                    ConflictRecord record = new ConflictRecord(context.getFileName(),
                                            context.getMod1Name(),
                                            context.getMod2Name(),
                                            signature,
                                            baseNode,
                                            modNode);
                                    record.setUserChoice(2);
                                    conflicts.add(record);
                                }
                                //当前节点不与原版相同，也不与待合并mod相同，标记为真正的冲突处
                                else {
                                    conflicts.add(new ConflictRecord(context.getFileName(),
                                            context.getMod1Name(),
                                            context.getMod2Name(),
                                            signature,
                                            baseNode,
                                            modNode));
                                }
                            }
                        }
                    }
                    //普通子节点，直接处理
                    else {
                        String baseText = baseNode.getSourceText();
                        String modText = modNode.getSourceText();
                        //内容不一致
                        if (!equalsTrimmed(baseText, modText)) {
                            // 检查modNode是否与原始基准MOD相同
                            if (!isNodeSameAsOriginalNode(originalNode, modNode)) {
                                conflicts.add(new ConflictRecord(context.getFileName(), context.getMod1Name(), context.getMod2Name(), signature, baseNode, modNode));
                            }
                        }
                    }
                }
            } catch (Exception e) {
                log.error("Error in processing scr node with signature: " + entry.getKey(), e);
            }
        }
    }

    private boolean isNodeSameAsOriginalNode(ScrScriptNode originalNode, ScrScriptNode modNode) {
        // 如果没有原始基准MOD，则认为不相同
        if (originalBasModRoot == null) {
            return false;
        }

        if (originalNode == null) {
            // 原始基准MOD中不存在这个节点，说明是新增的
            return false;
        }

        // 对比节点内容
        if (modNode instanceof ScrFunCallScriptNode modFunCall && originalNode instanceof ScrFunCallScriptNode originalFunCall) {
            // 函数调用节点，对比参数
            return modFunCall.getArguments().equals(originalFunCall.getArguments());
        } else {
            return equalsTrimmed(modNode.getSourceText(), originalNode.getSourceText());
        }
    }

    /**
     * 冲突解决
     */
    private void resolveConflictsInteractively() {
        //筛选出智能合并的节点
        List<ConflictRecord> automaticMerge = conflicts.stream()
                .filter(conflict -> conflict.getUserChoice() != null)
                .toList();
        if (!automaticMerge.isEmpty()) {
            for (ConflictRecord record : automaticMerge) {
                log.info("AutoMerging code line: {} -> {}", record.getBaseNode().getSourceText(), record.getModNode().getSourceText());
            }
            ColorPrinter.success(Localizations.t("SCR_MERGER_AUTO_MERGE_COUNT", automaticMerge.size()));
            conflicts.removeAll(automaticMerge);
        }
        if (conflicts.isEmpty()) {
            conflicts.addAll(automaticMerge); //智能合并的节点加回去，这里注意是为了不出现冲突提示
        } else {
            Scanner scanner = new Scanner(System.in);
            ColorPrinter.warning(Localizations.t("SCR_MERGER_CONFLICT_DETECTED", conflicts.size()));
            int chose = 0;
            for (int i = 0; i < conflicts.size(); i++) {
                ConflictRecord record = conflicts.get(i);
                if (chose == 3) {
                    record.setUserChoice(1); //3表示用户全部选择baseMod的配置来处理
                } else if (chose == 4) {
                    record.setUserChoice(2); //4表示用户全部选择mergeMod的配置来处理
                } else {
                    ColorPrinter.info("=".repeat(75));
                    ColorPrinter.info(Localizations.t("SCR_MERGER_FILE_INFO", i + 1, conflicts.size(), record.getFileName()));
                    //打印代码提示框
                    ColorPrinter.warning(Localizations.t("SCR_MERGER_MOD_VERSION_1", record.getBaseModName()));
                    ColorPrinter.bold(Localizations.t("SCR_MERGER_LINE_INFO", record.getBaseNode().getLine(), record.getBaseNode().getSourceText().trim()));
                    ColorPrinter.warning(Localizations.t("SCR_MERGER_MOD_VERSION_2", record.getMergeModName()));
                    ColorPrinter.bold(Localizations.t("SCR_MERGER_LINE_INFO", record.getModNode().getLine(), record.getModNode().getSourceText().trim()));
                    ColorPrinter.info("=".repeat(75));
                    //选择对话框
                    ColorPrinter.info(Localizations.t("SCR_MERGER_CHOOSE_PROMPT"));
                    ColorPrinter.info(Localizations.t("SCR_MERGER_USE_OPTION_1", record.getBaseNode().getSourceText()));
                    ColorPrinter.info(Localizations.t("SCR_MERGER_USE_OPTION_2", record.getModNode().getSourceText()));
                    ColorPrinter.info(Localizations.t("SCR_MERGER_USE_ALL_FROM_MOD_1", record.getBaseModName()));
                    ColorPrinter.info(Localizations.t("SCR_MERGER_USE_ALL_FROM_MOD_2", record.getMergeModName()));

                    while (true) {
                        String input = scanner.nextLine();
                        if (input.equals("1") || input.equals("2")) {
                            record.setUserChoice(Integer.parseInt(input));
                            break;
                        }
                        if (input.equals("3") || input.equals("4")) {
                            chose = Integer.parseInt(input);
                            record.setUserChoice(chose == 3 ? 1 : 2);
                            break;
                        }
                        ColorPrinter.warning(Localizations.t("SCR_MERGER_INVALID_INPUT"));
                    }
                }
            }
            conflicts.addAll(automaticMerge);  //智能合并的节点加回去，这里注意是为了不出现冲突提示
            ColorPrinter.success(Localizations.t("SCR_MERGER_CONFLICT_RESOLVED"));
        }
    }

    private void handleInsertion(ScrContainerScriptNode baseContainer, ScrScriptNode modNode) {
        // 插入位置：Base 容器的 '}' 之前
        int insertPos = baseContainer.getStopTokenIndex();
        String newContent = "\n    " + modNode.getSourceText();
        insertOperations.add(new InsertOperation(insertPos, newContent));
    }

    /**
     * 将MOD文件解析成语法树
     *
     * @param filePath 文件路径
     * @return 解析后的 ParseResult（包含AST和TokenStream）
     * @throws IOException 如果文件不可读
     */
    private static ParseResult parseFile(Path filePath) throws IOException {
        String content = Files.readString(filePath);
        String contentHash = Tools.computeHash(content);
        // 先查缓存
        ParseResult cached = PARSE_CACHE.get(contentHash);
        if (cached != null) {
            return cached;
        }
        // 缓存未命中，执行解析
        ParseResult result = parseContent(content);
        // 存入缓存
        PARSE_CACHE.put(contentHash, result);
        return result;
    }

    /**
     * 解析字符串内容为ParseResult
     */
    private static ParseResult parseContent(String content) {
        CharStream input = CharStreams.fromString(content);
        TechlandScriptLexer lexer = new TechlandScriptLexer(input);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        TechlandScriptParser parser = new TechlandScriptParser(tokens);
        TechlandScrFileVisitor visitor = new TechlandScrFileVisitor();
        // 注意：visitFile 返回的一定是我们定义的 ROOT Container
        ScrContainerScriptNode ast = (ScrContainerScriptNode) visitor.visitFile(parser.file());
        return new ParseResult(ast, tokens);
    }

    /**
     * 比较两段文本在忽略所有空白字符后的内容是否相同（避免创建新字符串）
     * <p>
     * 空白字符定义：Character.isWhitespace(c) 为 true 的字符（空格、Tab、换行等）。
     */
    private static boolean equalsTrimmed(String a, String b) {
        // 快路径：同一引用或 equals（完全一致时直接返回，避免扫描）
        if (Objects.equals(a, b)) {
            return true;
        }
        if (a == null || b == null) {
            return false;
        }

        int ai = 0;
        int bi = 0;
        int alen = a.length();
        int blen = b.length();

        while (true) {
            // 跳过 a 的所有空白
            while (ai < alen && Character.isWhitespace(a.charAt(ai))) {
                ai++;
            }
            // 跳过 b 的所有空白
            while (bi < blen && Character.isWhitespace(b.charAt(bi))) {
                bi++;
            }

            // 两边都到尾了 -> 相等
            if (ai >= alen || bi >= blen) {
                // 需要确保剩余部分也都是空白
                while (ai < alen && Character.isWhitespace(a.charAt(ai))) ai++;
                while (bi < blen && Character.isWhitespace(b.charAt(bi))) bi++;
                return ai >= alen && bi >= blen;
            }

            // 比较当前非空白字符
            if (a.charAt(ai) != b.charAt(bi)) {
                return false;
            }
            ai++;
            bi++;
        }
    }
}

File: src/main/java/ankol/mod/merger/merger/scr/TechlandScrFileMerger.kt
package ankol.mod.merger.merger.scr

import ankol.mod.merger.antlr.scr.TechlandScriptLexer
import ankol.mod.merger.antlr.scr.TechlandScriptParser
import ankol.mod.merger.constants.UserChoice
import ankol.mod.merger.core.BaseTreeNode
import ankol.mod.merger.core.ConflictResolver
import ankol.mod.merger.domain.MergerContext
import ankol.mod.merger.domain.ParsedResult
import ankol.mod.merger.core.filetrees.AbstractFileTree
import ankol.mod.merger.exception.BusinessException
import ankol.mod.merger.merger.AbstractFileMerger
import ankol.mod.merger.merger.ConflictRecord
import ankol.mod.merger.domain.MergeResult
import ankol.mod.merger.merger.scr.node.ScrContainerScriptNode
import ankol.mod.merger.merger.scr.node.ScrFunCallScriptNode
import ankol.mod.merger.tools.logger
import org.antlr.v4.runtime.CharStream
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.TokenStreamRewriter

class TechlandScrFileMerger(context: MergerContext) : AbstractFileMerger(context) {
    private val log = logger()

    /**
     * 标记冲突项的容器
     */
    private val conflicts = ArrayList<ConflictRecord>()

    /**
     * 插入操作记录
     */
    private data class InsertOperation(val tokenIndex: Int, val content: String, val nodeType: NodeType = NodeType.OTHER)

    private val insertOperations = ArrayList<InsertOperation>()

    /**
     * 节点类型，用于确定插入位置的优先级
     */
    private enum class NodeType {
        IMPORT,  // import 语句 - 最高优先级，放在文件最前
        SUB,     // sub 函数声明 - 次高优先级，放在 import 之后
        OTHER    // 其他声明 - 最低优先级
    }

    /**
     * 基准MOD（data0.pak）对应文件的语法树，用于三方对比
     */
    private var vanillaRootNode: ScrContainerScriptNode? = null

    override fun merge(accumulatedFile: AbstractFileTree, incomingModFile: AbstractFileTree): MergeResult {
        try {
            val parsedResult = context.baseModManager.parseForm(accumulatedFile.fileEntryName) { parseContent(it) }
            // 解析基准MOD文件（如果存在）
            if (parsedResult != null) {
                vanillaRootNode = parsedResult.astNode
            }
            // 解析base和mod文件，保留TokenStream
            val accumulatedResult = parseContent(accumulatedFile.getContent())
            val incomingModResult = parseContent(incomingModFile.getContent())
            val accumulatedRoot: ScrContainerScriptNode = accumulatedResult.astNode
            val incomingModRoot: ScrContainerScriptNode = incomingModResult.astNode

//            accumulatedRoot.printTree()
//            incomingModRoot.printTree()

            deepCompare(vanillaRootNode, accumulatedRoot, incomingModRoot)

            //第一个mod与原版文件的对比
            if (context.isFirstMerge && !conflicts.isEmpty()) {
                for (record in conflicts) {
                    record.userChoice = UserChoice.MERGE_MOD
                }
            } else if (!conflicts.isEmpty()) {
                // 正常情况下，提示用户解决冲突
                ConflictResolver.resolveConflict(conflicts, context)
            }

            return MergeResult(getMergedContent(accumulatedResult), context.mergedHistory)
        } catch (e: Exception) {
            log.error("Error during SCR file merge: ${accumulatedFile.fileName} Reason: ${e.message}", e)
            throw BusinessException("文件${accumulatedFile.fileName}合并失败")
        } finally {
            //清理状态，准备下一个文件合并
            conflicts.clear()
            insertOperations.clear()
            vanillaRootNode = null
        }
    }

    private fun deepCompare(
        vanillaContainer: ScrContainerScriptNode?,
        accumulatedContainer: ScrContainerScriptNode,
        incomingModContainer: ScrContainerScriptNode
    ) {
        // 遍历 Mod 的所有子节点
        for ((signature, incomingModNode) in incomingModContainer.childrens) {
            try {
                var vanillaNode: BaseTreeNode? = null
                if (vanillaContainer != null) {
                    vanillaNode = vanillaContainer.childrens[signature]
                }
                val accumulatedNode = accumulatedContainer.childrens[signature]

                if (accumulatedNode == null) {
                    handleInsertion(accumulatedContainer, incomingModNode)
                } else {
                    //容器节点，递归对比
                    if (accumulatedNode is ScrContainerScriptNode && incomingModNode is ScrContainerScriptNode) {
                        deepCompare(vanillaNode as ScrContainerScriptNode?, accumulatedNode, incomingModNode)
                    }
                    //function call节点的对比逻辑
                    else if (accumulatedNode is ScrFunCallScriptNode && incomingModNode is ScrFunCallScriptNode) {
                        if (accumulatedNode.arguments != incomingModNode.arguments) {
                            if (!isNodeSameAsOriginalNode(vanillaNode, incomingModNode)) {
                                if (isNodeSameAsOriginalNode(vanillaNode, accumulatedNode)) {
                                    context.mergedHistory.markSignture("${incomingModContainer.signature}-${signature}", context.mergeModName)
                                    //base节点与原版一致，自动合并
                                    val record = ConflictRecord(
                                        context.mergingFileName,
                                        context.accumulatedModName,
                                        context.mergeModName,
                                        signature,
                                        accumulatedNode,
                                        incomingModNode
                                    )
                                    record.userChoice = UserChoice.MERGE_MOD
                                    conflicts.add(record)
                                } else {
                                    //标记真正的冲突 todo MergedHistory存在作用域混乱的问题，还需改良
                                    val modName = context.mergedHistory.getModNameFromSignature("${incomingModContainer.signature}-${signature}")
                                    conflicts.add(
                                        ConflictRecord(
                                            context.mergingFileName,
                                            modName ?: context.accumulatedModName,
                                            context.mergeModName,
                                            signature,
                                            accumulatedNode,
                                            incomingModNode
                                        )
                                    )
                                }
                            }
                        }
                    }
                    //其他类型节点直接比较文本内容
                    else {
                        val accumulatedText = accumulatedNode.sourceText
                        val incomingModText = incomingModNode.sourceText
                        //内容不一致
                        if (!equalsTrimmed(accumulatedText, incomingModText)) {
                            if (!isNodeSameAsOriginalNode(vanillaNode, incomingModNode)) {
                                if (isNodeSameAsOriginalNode(vanillaNode, accumulatedNode)) {
                                    context.mergedHistory.markSignture(incomingModNode.signature, context.mergeModName)
                                    conflicts.add(
                                        ConflictRecord(
                                            context.mergingFileName,
                                            context.accumulatedModName,
                                            context.mergeModName,
                                            signature,
                                            accumulatedNode,
                                            incomingModNode,
                                            UserChoice.MERGE_MOD
                                        )
                                    )
                                } else {
                                    conflicts.add(
                                        ConflictRecord(
                                            context.mergingFileName,
                                            context.accumulatedModName,
                                            context.mergeModName,
                                            signature,
                                            accumulatedNode,
                                            incomingModNode
                                        )
                                    )
                                }
                            }
                        }
                    }
                }
            } catch (e: Exception) {
                log.error("Error in processing scr node with signature: '${signature}'", e)
            }
        }
    }

    private fun getMergedContent(accumulatedResult: ParsedResult<ScrContainerScriptNode>): String {
        val rewriter = TokenStreamRewriter(accumulatedResult.tokenStream)
        // 处理冲突节点的替换
        for (record in conflicts) {
            if (record.userChoice == UserChoice.MERGE_MOD) {
                // 普通修改冲突：用户选择了 Mod
                val accumulatedNode = record.baseNode
                val incomingModNode = record.modNode
                rewriter.replace(accumulatedNode.startTokenIndex, accumulatedNode.stopTokenIndex, incomingModNode.sourceText)
            }
        }

        // 对插入操作按照优先级和位置排序
        // 优先级：IMPORT > SUB > OTHER
        // 同一优先级内按照 tokenIndex 升序排序（从前往后插入）
        val sortedOperations = insertOperations.sortedWith(compareBy<InsertOperation> { op ->
            when (op.nodeType) {
                NodeType.IMPORT -> 0
                NodeType.SUB -> 1
                NodeType.OTHER -> 2
            }
        }.thenBy { it.tokenIndex })

        for (op in sortedOperations) {
            rewriter.insertBefore(op.tokenIndex, op.content)
        }

        // 获取重写后的文本
        return rewriter.text
    }

    private fun isNodeSameAsOriginalNode(originalNode: BaseTreeNode?, incomingModNode: BaseTreeNode): Boolean {
        // 如果没有原始基准MOD，则认为不相同
        if (vanillaRootNode == null) {
            return false
        }

        if (originalNode == null) {
            // 原始基准MOD中不存在这个节点，说明是新增的
            return false
        }

        // 对比节点内容
        return if (incomingModNode is ScrFunCallScriptNode && originalNode is ScrFunCallScriptNode) {
            // 函数调用节点，对比参数
            incomingModNode.arguments == originalNode.arguments
        } else {
            equalsTrimmed(incomingModNode.sourceText, originalNode.sourceText)
        }
    }

    private fun handleInsertion(accumulatedContainer: ScrContainerScriptNode, incomingModNode: BaseTreeNode) {
        // 根据节点签名确定节点类型
        val nodeType = when {
            incomingModNode.signature.startsWith("import:") -> NodeType.IMPORT
            incomingModNode.signature.startsWith("sub:") -> NodeType.SUB
            else -> NodeType.OTHER
        }

        var newContent: String

        // 选择合适的插入位置
        val insertPos = when (nodeType) {
            NodeType.IMPORT -> {
                // import语句需要插入在文件最顶上
                newContent = "\n${incomingModNode.sourceText}"
                findInsertPositionForImport(accumulatedContainer)
            }

            NodeType.SUB -> {
                // sub 插入到import语句后，但在其他节点前
                newContent = "${incomingModNode.sourceText}\n"
                findInsertPositionForSub(accumulatedContainer)
            }

            NodeType.OTHER -> {
                // 其他节点直接插在容器的 '}' 之前
                newContent = "\n   ${incomingModNode.sourceText}"
                accumulatedContainer.stopTokenIndex
            }
        }

        insertOperations.add(InsertOperation(insertPos, newContent, nodeType))
    }

    /**
     * 找到import语句的插入位置：在所有现有import之后，或者在第一个非import节点之前
     */
    private fun findInsertPositionForImport(container: ScrContainerScriptNode): Int {
        var lastImportStopIndex: Int? = null

        for ((_, node) in container.childrens) {
            if (node.signature.startsWith("import:")) {
                lastImportStopIndex = node.stopTokenIndex
            } else {
                // 遇到第一个非import节点，如果有import就插在最后import之后，否则插在该节点之前
                return lastImportStopIndex?.let { it + 1 } ?: node.startTokenIndex
            }
        }

        // 所有节点都是import，或者没有节点，返回容器结束位置
        return lastImportStopIndex?.let { it + 1 } ?: container.stopTokenIndex
    }

    /**
     * 找到sub函数的插入位置：在所有import和sub之后，但在其他节点之前
     */
    private fun findInsertPositionForSub(container: ScrContainerScriptNode): Int {
        var lastSubOrImportStopIndex: Int? = null

        for ((_, node) in container.childrens) {
            val isSub = node.signature.startsWith("sub:")
            val isImport = node.signature.startsWith("import:")

            if (isSub || isImport) {
                lastSubOrImportStopIndex = node.startTokenIndex
            } else {
                // 遇到第一个既不是import也不是sub的节点
                return lastSubOrImportStopIndex?.let { it + 1 } ?: node.startTokenIndex
            }
        }

        // 所有节点都是import或sub，返回容器结束位置
        return lastSubOrImportStopIndex ?: container.stopTokenIndex
    }

    private fun parseContent(content: String): ParsedResult<ScrContainerScriptNode> {
        val input: CharStream = CharStreams.fromString(content)
        val lexer = TechlandScriptLexer(input)
        val tokens = CommonTokenStream(lexer)
        val parser = TechlandScriptParser(tokens)
        val visitor = TechlandScrFileVisitor(tokens)
        // 注意：visitFile 返回的一定是我们定义的 ROOT Container
        val ast = visitor.visitFile(parser.file()) as ScrContainerScriptNode
        return ParsedResult(ast, tokens)
    }

    private fun equalsTrimmed(a: String?, b: String?): Boolean {
        // 快路径：同一引用或 equals（完全一致时直接返回，避免扫描）
        if (a == b) {
            return true
        }
        if (a == null || b == null) {
            return false
        }

        var ai = 0
        var bi = 0
        val alen = a.length
        val blen = b.length

        while (true) {
            // 跳过 a 的所有空白
            while (ai < alen && Character.isWhitespace(a[ai])) {
                ai++
            }
            // 跳过 b 的所有空白
            while (bi < blen && Character.isWhitespace(b[bi])) {
                bi++
            }

            // 两边都到尾了 -> 相等
            if (ai >= alen || bi >= blen) {
                // 需要确保剩余部分也都是空白
                while (ai < alen && Character.isWhitespace(a[ai])) ai++
                while (bi < blen && Character.isWhitespace(b[bi])) bi++
                return ai >= alen && bi >= blen
            }

            // 比较当前非空白字符
            if (a[ai] != b[bi]) {
                return false
            }
            ai++
            bi++
        }
    }
}
